File: frontend/src/api/person.ts
import api from './index'

// 人员类型定义
export interface PersonBase {
  id: string
  name: string
  gender: number
  birthday?: string
  phone?: string
  email?: string
  type: 'student' | 'teacher' | 'parent'
}

// 老师班级关联创建结构
export interface TeacherClassCreate {
  class_id: string
  is_main_teacher: boolean
}

// 老师班级关联响应结构
export interface TeacherClassResponse {
  id: string
  name: string
  grade: number
  academic_year: string
  is_main_teacher: boolean
}

export interface StudentResponse extends PersonBase {
  type: 'student'
  student_no: string
  class_id?: string
  class_name?: string
  enrollment_date?: string
  status: string
}

export interface TeacherResponse extends PersonBase {
  type: 'teacher'
  employee_no: string
  department_id?: string
  department_name?: string
  title?: string
  hire_date?: string
  // 老师关联的多个班级
  classes?: TeacherClassCreate[]
}

export interface ParentResponse extends PersonBase {
  type: 'parent'
  wechat_openid?: string
  occupation?: string
}

export type PersonResponse = StudentResponse | TeacherResponse | ParentResponse

export interface PersonCreate {
  name: string
  gender: number
  birthday?: string
  phone?: string
  email?: string
  type_: string
  student_no?: string
  class_id?: string
  enrollment_date?: string
  employee_no?: string
  department_id?: string
  title?: string
  hire_date?: string
  wechat_openid?: string
  occupation?: string
  // 老师关联的多个班级
  classes?: TeacherClassCreate[]
}

export interface PersonUpdate {
  name?: string
  gender?: number
  birthday?: string
  phone?: string
  email?: string
  student_no?: string
  class_id?: string
  enrollment_date?: string
  employee_no?: string
  department_id?: string
  title?: string
  hire_date?: string
  wechat_openid?: string
  occupation?: string
  // 老师关联的多个班级
  classes?: TeacherClassCreate[]
}

export interface PersonQuery {
  page: number
  limit: number
  type?: string
  search?: string
  class_id?: string
  department_id?: string
}

export interface ListResponse<T> {
  items: T[]
  total: number
  page: number
  limit: number
}

// 人员管理API
export const personApi = {
  // 获取人员列表
  list: (params: PersonQuery) => {
    return api.get<ListResponse<PersonResponse>>('/persons', { params })
  },
  
  // 创建人员
  create: (data: PersonCreate) => {
    return api.post<PersonResponse>('/persons', data)
  },
  
  // 获取单个人员
  get: (id: string) => {
    return api.get<PersonResponse>(`/persons/${id}`)
  },
  
  // 更新人员
  update: (id: string, data: PersonUpdate) => {
    return api.put<PersonResponse>(`/persons/${id}`, data)
  },
  
  // 删除人员
  delete: (id: string) => {
    return api.delete(`/persons/${id}`)
  },
  
  // 获取老师关联的班级
  getTeacherClasses: (teacherId: string) => {
    return api.get<TeacherClassResponse[]>('/permission/teacher/classes', { params: { teacher_id: teacherId } })
  }
}

File: backend/src/models/person.rs
use chrono::{NaiveDate, DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Deserialize, Serialize, sqlx::FromRow)]
pub struct Person {
    pub id: Uuid,
    pub name: String,
    pub gender: i16, // 0:未知 1:男 2:女
    pub birthday: Option<NaiveDate>,
    pub phone: Option<String>,
    pub email: Option<String>,
    #[sqlx(rename = "type")]
    pub type_: String, // 'student', 'teacher', 'parent'
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct PersonCreate {
    pub name: String,
    pub gender: i32,
    pub birthday: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub type_: String,
    // 子类型特定字段，根据type_决定哪些字段有效
    pub student_no: Option<String>,
    pub class_id: Option<Uuid>,
    pub enrollment_date: Option<String>,
    pub employee_no: Option<String>,
    pub department_id: Option<Uuid>,
    pub title: Option<String>,
    pub hire_date: Option<String>,
    pub wechat_openid: Option<String>,
    pub occupation: Option<String>,
    // 老师关联的多个班级
    pub classes: Option<Vec<TeacherClassCreate>>,
}

// 老师班级关联创建结构
#[derive(Debug, Deserialize, Serialize)]
pub struct TeacherClassCreate {
    pub class_id: Uuid,
    pub is_main_teacher: bool,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct PersonUpdate {
    pub name: Option<String>,
    pub gender: Option<i32>,
    pub birthday: Option<String>,
    pub phone: Option<String>,
    pub email: Option<String>,
    // 子类型特定字段
    pub student_no: Option<String>,
    pub class_id: Option<Uuid>,
    pub enrollment_date: Option<String>,
    pub employee_no: Option<String>,
    pub department_id: Option<Uuid>,
    pub title: Option<String>,
    pub hire_date: Option<String>,
    pub wechat_openid: Option<String>,
    pub occupation: Option<String>,
    // 老师关联的多个班级
    pub classes: Option<Vec<TeacherClassCreate>>,
}

// 子类型结构体（用于查询连接结果）
#[derive(Debug, Deserialize, Serialize, sqlx::FromRow)]
pub struct Student {
    pub person_id: Uuid,
    pub student_no: String,
    pub class_id: Option<Uuid>,
    pub enrollment_date: Option<NaiveDate>,
    pub status: String,
}

#[derive(Debug, Deserialize, Serialize, sqlx::FromRow)]
pub struct Teacher {
    pub person_id: Uuid,
    pub employee_no: String,
    pub department_id: Option<Uuid>,
    pub title: Option<String>,
    pub hire_date: Option<NaiveDate>,
}

// 老师与班级的关联关系
#[derive(Debug, Deserialize, Serialize, sqlx::FromRow)]
pub struct TeacherClass {
    pub teacher_id: Uuid,
    pub class_id: Uuid,
    pub is_main_teacher: bool,
}

#[derive(Debug, Deserialize, Serialize, sqlx::FromRow)]
pub struct Parent {
    pub person_id: Uuid,
    pub wechat_openid: Option<String>,
    pub occupation: Option<String>,
}

// 统一响应枚举
#[derive(Debug, Deserialize, Serialize)]
#[serde(tag = "type")]
pub enum PersonResponse {
    #[serde(rename = "student")]
    Student(StudentResponse),
    #[serde(rename = "teacher")]
    Teacher(TeacherResponse),
    #[serde(rename = "parent")]
    Parent(ParentResponse),
}

#[derive(Debug, Deserialize, Serialize)]
pub struct StudentResponse {
    pub id: Uuid,
    pub name: String,
    pub gender: i16,
    pub birthday: Option<NaiveDate>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub student_no: String,
    pub class_id: Option<Uuid>,
    pub class_name: Option<String>,
    pub enrollment_date: Option<NaiveDate>,
    pub status: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TeacherClassInfo {
    pub class_id: Uuid,
    pub class_name: String,
    pub is_main_teacher: bool,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TeacherResponse {
    pub id: Uuid,
    pub name: String,
    pub gender: i16,
    pub birthday: Option<NaiveDate>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub employee_no: String,
    pub department_id: Option<Uuid>,
    pub department_name: Option<String>,
    pub classes: Vec<TeacherClassInfo>,
    pub title: Option<String>,
    pub hire_date: Option<NaiveDate>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ParentResponse {
    pub id: Uuid,
    pub name: String,
    pub gender: i16,
    pub birthday: Option<NaiveDate>,
    pub phone: Option<String>,
    pub email: Option<String>,
    pub wechat_openid: Option<String>,
    pub occupation: Option<String>,
}

// 转换实现
impl From<(Person, Option<Student>)> for PersonResponse {
    fn from((person, student): (Person, Option<Student>)) -> Self {
        match person.type_.as_str() {
            "student" if student.is_some() => {
                let student = student.unwrap();
                PersonResponse::Student(StudentResponse {
                    id: person.id,
                    name: person.name,
                    gender: person.gender,
                    birthday: person.birthday,
                    phone: person.phone,
                    email: person.email,
                    student_no: student.student_no,
                    class_id: student.class_id,
                    class_name: None, // 需要额外查询
                    enrollment_date: student.enrollment_date,
                    status: student.status
                })
            }
            "teacher" => {
                // 需要教师信息，暂时返回基础信息
                PersonResponse::Teacher(TeacherResponse {
                    id: person.id,
                    name: person.name,
                    gender: person.gender,
                    birthday: person.birthday,
                    phone: person.phone,
                    email: person.email,
                    employee_no: String::new(),
                    department_id: None,
                    department_name: None,
                    classes: Vec::new(),
                    title: None,
                    hire_date: None
                })
            }
            "parent" => {
                PersonResponse::Parent(ParentResponse {
                    id: person.id,
                    name: person.name,
                    gender: person.gender,
                    birthday: person.birthday,
                    phone: person.phone,
                    email: person.email,
                    wechat_openid: None,
                    occupation: None
                })
            }
            _ => {
                // 默认作为基础人员
                PersonResponse::Teacher(TeacherResponse {
                    id: person.id,
                    name: person.name,
                    gender: person.gender,
                    birthday: person.birthday,
                    phone: person.phone,
                    email: person.email,
                    employee_no: String::new(),
                    department_id: None,
                    department_name: None,
                    classes: Vec::new(),
                    title: None,
                    hire_date: None
                })
            }
        }
    }
}
